x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx

```diff
@@ -0,0 +1,204 @@
+    // @ts-ignore - TODO remove this ignore when support for TransactionV1 is dropped
+    const podId = oc(transaction).kubernetes.pod.uid();
+    // @ts-ignore - TODO remove this ignore when support for TransactionV1 is dropped
+    const containerId = oc(transaction).container.id();
+
+    const pathname = '/app/infra';
+    const time = new Date(transaction['@timestamp']).getTime();
+    const infraMetricsQuery = getInfraMetricsQuery(transaction);
+
+    return [
+      {
+        icon: 'loggingApp',
+        label: 'Show trace logs',
+        target: traceId,
+        hash: `/link-to/host-logs/${hostName}`,
+        query: { time, traceId }
+      },
+
+      {
+        icon: 'loggingApp',
+        label: 'Show pod logs',
+        target: podId,
+        hash: `/link-to/pod-logs/${podId}`,
+        query: { time }
+      },
+
+      {
+        icon: 'loggingApp',
+        label: 'Show container logs',
+        target: containerId,
+        hash: `/link-to/container-logs/${containerId}`,
+        query: { time }
+      },
+
+      {
+        icon: 'loggingApp',
+        label: 'Show host logs',
+        target: hostName,
+        hash: `/link-to/host-logs/${hostName}`,
+        query: { time }
+      },
+
+      {
+        icon: 'infraApp',
+        label: 'Show pod metrics',
+        target: podId,
+        hash: `/link-to/pod-detail/${podId}`,
+        query: infraMetricsQuery
+      },
+
+      {
+        icon: 'infraApp',
+        label: 'Show container metrics',
+        target: containerId,
+        hash: `/link-to/container-detail/${containerId}`,
+        query: infraMetricsQuery
+      },
+
+      {
+        icon: 'infraApp',
+        label: 'Show host metrics',
+        target: hostName,
+        hash: `/link-to/host-detail/${hostName}`,
+        query: infraMetricsQuery
+      }
+    ]
+      .filter(({ target }) => Boolean(target))
+      .map(({ icon, label, hash, query }, index) => {
+        const href = getUnconnectedKibanaHref({
+          location,
+          pathname,
+          hash,
+          query
+        });
+
+        return (
+          <EuiContextMenuItem icon={icon} href={href} key={index}>
+            <EuiFlexGroup gutterSize="s">
+              <EuiFlexItem>
+                <EuiLink>{label}</EuiLink>
+              </EuiFlexItem>
+              <EuiFlexItem grow={false}>
+                <EuiIcon type="popout" />
+              </EuiFlexItem>
+            </EuiFlexGroup>
+          </EuiContextMenuItem>
+        );
+      });
+  }
+
+  public render() {
+    const { transaction, location } = this.props;
+
+    const discoverTransactionHref = getUnconnectedKibanaHref({
+      location,
+      pathname: '/app/kibana',
+      hash: '/discover',
+      query: getDiscoverQuery(transaction)
+    });
+
+    const items = [
+      ...this.getInfraActions(transaction),
+      <EuiContextMenuItem
+        icon="discoverApp"
+        href={discoverTransactionHref}
+        key="discover-transaction"
+      >
```

Comment:
You also need to specify the index pattern like in https://github.com/elastic/kibana/blob/26c77eb25edd5129265208c24dd3a6ab68a2885b/x-pack/plugins/apm/public/components/shared/DiscoverButtons/DiscoverButton.tsx#L34-L36

Instead of duplicating that, perhaps you can make some share method for doing this.

Note: `getAPMIndexPattern` calls a memoized function, so there is no need to avoid additional calls.

Reply:
good catch. I abstracted this logic into an async function `setApmIndexPatternQuery(query)` and use it in both places